frontend/src/app/services/auth.service.ts

```diff
@@ -1,153 +1,166 @@
 import { Injectable } from '@angular/core';
 import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
 import { SessionService } from './session.service';
-import { Router } from '@angular/router';
-import { BehaviorSubject, Observable, catchError, map, mergeMap, of, tap } from 'rxjs';
-import { User, UserResponse } from '../user/models/user';
+import { Subject, catchError, map, of, switchMap, merge, share, startWith, withLatestFrom, shareReplay, throttleTime, debounceTime } from 'rxjs';
+import { UserRegistration, UserResponse, UserLogin, PasswordForgotten, ResetPassword, KeyInfo, UserSettings } from '../user/models/user';
 import { encodeUserData, parseUserData } from '../user/utils';
 import _ from 'underscore';
 import { HttpClient } from '@angular/common/http';
 
+interface AuthAPIResult {
+    detail: string;
+}
+
+interface AuthAPIError {
+    error: Record<string, string>;
+}
+
 @Injectable({
     providedIn: 'root'
 })
 export class AuthService {
-    private currentUserSubject$ = new BehaviorSubject<User | null | undefined>(undefined);
+    public login$ = new Subject<UserLogin>();
+    public loginResult$ = this.login$.pipe(
+        throttleTime(500),
+        switchMap(loginForm => this.http.post<AuthAPIResult>(
+            this.authRoute('login/'), loginForm
+        ).pipe(
+            catchError(error => of<AuthAPIError>({ error: error.error })),
```

Comment:
I agree that there are quite a lot of / too many `catchError` operators here. The `error` callback sounds good, but it requires subscribing, which is otherwise not always necessary. I'm basically using `catchError` as a quasi-`mapToError` operator, based on https://www.learnrxjs.io/learn-rxjs/operators/transformation/partition (but without using the `partition` operator itself). If you have an alternative way of obtaining a stream of errors, I'd be happy to drop `catchError`.

Once I've added the util that creates a subject + 3 streams using a class, as per your suggestion, it will hopefully look more maintainable, but I'm interested to hear what you think.